src/sentry/eventstream/kafka/consumer.py

```diff
@@ -0,0 +1,268 @@
+            # consumer has received its assignment.)
+            return
+
+        # TODO: This wil be called from the commit log consumer thread, so need
+        # to verify that calling the ``consumer.{pause,resume}`` methods is
+        # thread safe!
+        if current_state in (SynchronizedPartitionState.UNKNOWN, SynchronizedPartitionState.SYNCHRONIZED,
+                             SynchronizedPartitionState.REMOTE_BEHIND):
+            self.__consumer.pause([TopicPartition(topic, partition, current_offsets.local)])
+        elif current_state is SynchronizedPartitionState.LOCAL_BEHIND:
+            self.__consumer.resume([TopicPartition(topic, partition, current_offsets.local)])
+        else:
+            raise NotImplementedError('Unexpected partition state: %s' % (current_state,))
+
+    def __get_initial_offset(self, topic, partition):
+        low, high = self.__consumer.get_watermark_offsets(TopicPartition(topic, partition))
+        return low
+
+    def subscribe(self, topics, on_assign=None, on_revoke=None):
+        """
+        Subscribe to a topic.
+        """
+        def assignment_callback(consumer, assignment):
+            # Since ``auto.offset.reset`` is set to ``error`` to force human
+            # interaction on an offset reset, we have to explicitly specify the
+            # starting offset if no offset has been committed for this topic during
+            # the ``__consumer_offsets`` topic retention period.
+            assignment = {
+                (i.topic, i.partition): self.__positions.get((i.topic, i.partition)) for i in assignment
+            }
+
+            for i in self.__consumer.committed([TopicPartition(topic, partition) for (
+                    topic, partition), offset in assignment.items() if offset is None]):
+                k = (i.topic, i.partition)
+                if i.offset > -1:
+                    assignment[k] = i.offset
+                else:
+                    assignment[k] = self.__get_initial_offset(i.topic, i.partition)
+
+            self.__consumer.assign([TopicPartition(topic, partition, offset)
+                                    for (topic, partition), offset in assignment.items()])
+
+            for (topic, partition), offset in assignment.items():
+                # Setting the local offsets will either cause the partition to be
+                # paused (if the remote offset is unknown or the local offset is
+                # not trailing the remote offset) or resumed.
+                self.__partition_state_manager.set_local_offset(topic, partition, offset)
+                self.__positions[(topic, partition)] = offset
+
+            if on_assign is not None:
+                on_assign(self, [TopicPartition(topic, partition)
+                                 for topic, partition in assignment.keys()])
+
+        def revocation_callback(consumer, assignment):
+            if on_revoke is not None:
+                on_revoke(self, assignment)
```

Comment:
Same note about callbacks as in the constructor

Reply:
Seems harmless.